TestKit Documentation (#139)
akka/akka-projection: docs/src/main/paradox/testing.md

# Testing

Akka Projection provides a Test Kit to ease testing. There are two supported styles of test: running with an assert function and driving it with a Akka Streams TestKit `SinkProbe`.

## Dependencies

To use the Akka Projection TestKit add the following dependency in your project:

@@dependency [sbt,Maven,Gradle] {
  group=com.typesafe.akka
  artifact=akka-projection-testkit_$scala.binary.version$
  version=$project.version$
  scope="test"
}

Akka Projections require Akka $akka.version$ or later, see @ref:[Akka version](overview.md#akka-version).

@@project-info{ projectId="akka-projection-testkit" }

### Transitive dependencies

> The table below shows `akka-projection-cassandra`'s direct dependencies and the second tab shows all libraries it depends on transitively.

Reviewer: akka-projection-testkit

# Testing

> Akka Projection provides a Test Kit to ease testing. There are two supported styles of test: running with an assert function and driving it with a Akka Streams TestKit `TestSink`.

Reviewer 2: ```suggestion
Akka Projection provides a Test Kit to ease testing. There are two supported styles of test: running with an assert function and driving it with an Akka Streams TestKit `TestSink`.
```

  version=$project.version$
  scope="test"
}

Akka Projections require Akka $akka.version$ or later, see @ref:[Akka version](overview.md#akka-version).

@@project-info{ projectId="akka-projection-testkit" }

### Transitive dependencies

The table below shows `akka-projection-testkit`'s direct dependencies and the second tab shows all libraries it depends on transitively.

@@dependencies{ projectId="akka-projection-testkit" }

## Initializing the Projection Test Kit

The Projection Test Kit requires an instance of `ActorTestKit`. We recommend using Akka's @scala[`ScalaTestWithActorTestKit`]@java[`TestKitJunitResource`]

Scala
:  @@snip [TestKitDocExample.scala](/examples/src/test/scala/docs/testkit/TestKitDocExample.scala) { #testkit-import #testkit }

Java
:  @@snip [TestKitDocExample.java](/examples/src/test/java/jdocs/testkit/TestKitDocExample.java) { #testkit-import #testkit }

## Testing with an assert function

When testing with an assert function the Projection is started and stopped by the test kit. While the projection is running, the assert function will be called until it completes without errors (no exceptions or assertion errors are thrown). 

> In the example bellow the Projection will update a Shopping Cart Checkout view. The test will run until it observes that the shopping cart for id `abc-def` was checkout.  

Reviewer 2: ```suggestion
In the example below the Projection will update a Shopping Cart Checkout view. The test will run until it observes that the shopping cart for id `abc-def` was checked out.  
```

Scala
:  @@snip [TestKitDocExample.scala](/examples/src/test/scala/docs/testkit/TestKitDocExample.scala) { #testkit-import #testkit-run }

Java
:  @@snip [TestKitDocExample.java](/examples/src/test/java/jdocs/testkit/TestKitDocExample.java) { #fixme }

By default, the test will run for 3 seconds. The assert function will be called every 100 milliseconds. Those values can be modified programatically.

Scala
:  @@snip [TestKitDocExample.scala](/examples/src/test/scala/docs/testkit/TestKitDocExample.scala) { #testkit-duration #testkit-run-max-interval }

Java
:  @@snip [TestKitDocExample.java](/examples/src/test/java/jdocs/testkit/TestKitDocExample.java) { #fixme }  

## Testing with a TestSink

The [Akka Stream Test Kit](https://doc.akka.io/docs/akka/current/stream/stream-testkit.html#using-the-testkit) can be used to drive the pace of envelopes flowing through the Projection.   

Scala
:  @@snip [TestKitDocExample.scala](/examples/src/test/scala/docs/testkit/TestKitDocExample.scala) { #testkit-duration #testkit-sink-probe }

Java
:  @@snip [TestKitDocExample.java](/examples/src/test/java/jdocs/testkit/TestKitDocExample.java) { #fixme }

> Differently then testing with assert function  

Reviewer 2: ```suggestion
Differently than testing with assert function  
```

Incomplete sentence.
